File: backend/src/dto/proof_matrix.rs
//! Response DTOs for `GET /api/cases/:slug/proof-matrix/rollup` — the first
//! piece of the Proof Matrix compute layer.
//!
//! This endpoint is the single source of truth for Count-level *deduped*
//! allegation totals: per `LegalCount`, the number of DISTINCT `Allegation`s
//! that bear on any of that Count's `Element`s. (Home will consume this later;
//! it is not wired here.)
//!
//! Serialize-only response types — these are produced by the backend and sent
//! to the client; nothing here is decoded from stored JSON, so (unlike the
//! causes-of-action DTOs) no `Deserialize` is needed.
//!
//! Field names are snake_case and align 1:1 with the Cypher result aliases
//! (`count_number`, `count_id`, `deduped_allegations`) so the wire shape is
//! traceable straight back to the query.

use serde::Serialize;

/// Top-level payload: the requested case slug (echoed) and the per-Count
/// rollup rows.
///
/// ## Rust Learning: `#[derive(Serialize)]` without `Deserialize`
///
/// `serde::Serialize` is the only half of the serde pair we need here: axum's
/// `Json(...)` wrapper turns this struct into the HTTP response body. We do not
/// derive `Deserialize` because the backend never parses this shape back in —
/// deriving only what is used keeps the type honest about its direction of flow
/// and avoids an unused trait impl.
#[derive(Debug, Clone, Serialize)]
pub struct ProofMatrixRollupResponse {
    /// Echoed from the request path. The Neo4j graph is single-case and not
    /// slug-namespaced, so the slug does not filter the query — it is returned
    /// for the caller's correlation, matching the causes-of-action contract.
    pub case_slug: String,
    /// One entry per `LegalCount` that has at least one bearing Allegation,
    /// ordered by `count_number` ascending (the query's `ORDER BY`).
    pub counts: Vec<CountRollup>,
}

/// One Count's deduped allegation total.
///
/// Domain note: `deduped_allegations` is `count(DISTINCT a)` *per Count*. An
/// Allegation that bears on several Elements of the same Count is counted once
/// here. This is deliberately NOT the same as summing the per-Element
/// allegation counts exposed by the causes-of-action endpoint, where that same
/// Allegation is counted once per Element it touches. See
/// `proof_matrix_repository` for the full reconciliation note.
#[derive(Debug, Clone, Serialize)]
pub struct CountRollup {
    /// The Count's ordinal (1-4 in the current case), from `LegalCount.count_number`.
    pub count_number: i64,
    /// The Count's stable identifier, from `LegalCount.id` (e.g. `"count-1"`).
    pub count_id: String,
    /// Number of DISTINCT Allegations bearing on any Element of this Count.
    pub deduped_allegations: i64,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    /// Pins the wire contract: top-level `case_slug` + `counts`, and each row's
    /// exact snake_case field names and value types. A rename of any field (or
    /// an accidental `camelCase` serde attribute) would break Home's later
    /// wiring silently — this catches it at `cargo test` time.
    #[test]
    fn rollup_response_serializes_with_expected_field_names() {
        let response = ProofMatrixRollupResponse {
            case_slug: "awad_v_catholic_family_service".to_string(),
            counts: vec![
                CountRollup {
                    count_number: 1,
                    count_id: "count-1".to_string(),
                    deduped_allegations: 51,
                },
                CountRollup {
                    count_number: 2,
                    count_id: "count-2".to_string(),
                    deduped_allegations: 41,
                },
            ],
        };

        let value = serde_json::to_value(&response).expect("response serializes");
        assert_eq!(
            value,
            json!({
                "case_slug": "awad_v_catholic_family_service",
                "counts": [
                    {"count_number": 1, "count_id": "count-1", "deduped_allegations": 51},
                    {"count_number": 2, "count_id": "count-2", "deduped_allegations": 41}
                ]
            })
        );
    }

    /// An empty graph (no bearing Allegations on any Count) must serialize as an
    /// empty array, never `null` — "no rows" stays distinguishable from "field
    /// absent" (Standing Rule 1). The handler turns this case into a 404 before
    /// it reaches serialization, but the DTO must still encode it correctly.
    #[test]
    fn empty_counts_serialize_as_array_not_null() {
        let response = ProofMatrixRollupResponse {
            case_slug: "awad_v_catholic_family_service".to_string(),
            counts: vec![],
        };
        let value = serde_json::to_value(&response).expect("response serializes");
        assert_eq!(
            value,
            json!({"case_slug": "awad_v_catholic_family_service", "counts": []})
        );
    }
}

File: frontend/src/services/proofMatrix.ts
// =============================================================================
// proofMatrix.ts — client for GET /api/cases/:slug/proof-matrix/rollup
// -----------------------------------------------------------------------------
// Endpoint: GET /api/cases/:slug/proof-matrix/rollup → backend handler
// `api::proof_matrix`. Neo4j-backed, read-only. Returns, per LegalCount, the
// count of DISTINCT Allegations bearing on that Count's Elements — the single
// source of truth for Count-level deduped allegation totals.
//
// These interfaces mirror the Rust DTO (`dto::proof_matrix`) exactly. This is a
// SEPARATE endpoint from causes-of-action: the deduped per-Count total is NOT
// the same number as summing the per-Element `allegation_count` fields that
// causes-of-action returns (an Allegation bearing on two Elements of one Count
// is counted once here, twice there). We therefore keep this in its own service
// and its own type — never merged onto `CountDetail`, which mirrors a different
// wire DTO.
// =============================================================================

import { API_BASE_URL } from "./api";
import { authFetch } from "./auth";
import { DEFAULT_CASE_SLUG } from "./caseHeader";

/** One Count's deduped allegation total. `count_number` is the join key. */
export type CountRollup = {
  count_number: number;
  count_id: string;
  deduped_allegations: number;
};

/** Top-level payload: the echoed slug and the per-Count rollup rows. */
export type ProofMatrixRollupResponse = {
  case_slug: string;
  counts: CountRollup[];
};

// ─── Structural-column types (PM4) ──────────────────────────────────────────
//
// `EvidenceRef` used to live here: a placeholder shape for "what the evidence
// columns will hold once discovery is processed", declared while both columns
// were fed empty arrays. It was deleted on 2026-07-27 with the last thing that
// rendered it. The anticipated shape never became the real one — the Supporting
// and Disputes columns show backend-computed COUNTS, and the items behind them
// come from the element-detail endpoint as `AllegationEvidence`
// (services/elementDetailService.ts), a different shape carrying the source
// locator the placeholder lacked.
//
// Worth remembering rather than just deleting: the placeholder outlived its
// premise. It was defined on the assumption that zero Evidence nodes existed,
// and stayed after that stopped being true — which is why the Disputes column
// sat blank over 41 real REBUTS edges. Shapes designed ahead of their data need
// a reason to be revisited, not just a comment saying Stage 2 will swap them.

/**
 * Proof status for one Element — the EXACT lowercase values the backend's
 * `derive_proof_status` emits (Part 2 aligned these to the wire). Computed by
 * the backend from coverage; the frontend only renders it (Rule 19 — no
 * client-side status derivation).
 *
 * Domain note: this is **presence-of-evidence**, NOT legal sufficiency — there
 * is intentionally no `"proven"`. `"no_allegations"` (nothing mapped to the
 * Element) is distinct from `"gap"` (allegations mapped, none corroborated).
 */
export type ElementProofStatus =
  | "no_allegations"
  | "gap"
  | "partial"
  | "supported";

/**
 * Fetch the per-Count deduped allegation rollup for `slug` (defaults to the
 * single seeded case).
 *
 * Mirrors `getCausesOfAction`: validates the load-bearing field (`counts` is an
 * array) and throws a contextual error at the boundary rather than letting a
 * malformed body crash a component later — Standing Rule 1 (no silent failures).
 *
 * @param slug case slug; defaults to {@link DEFAULT_CASE_SLUG}
 * @returns the typed proof-matrix rollup payload
 * @throws Error on non-2xx, unparseable body, or a body missing `counts`
 */
export async function getProofMatrixRollup(
  slug: string = DEFAULT_CASE_SLUG,
): Promise<ProofMatrixRollupResponse> {
  // authFetch adds credentials + a 30s timeout (AbortController) — Rule 13.
  const response = await authFetch(
    `${API_BASE_URL}/api/cases/${encodeURIComponent(slug)}/proof-matrix/rollup`,
  );

  if (!response.ok) {
    // 404 here means the canonical case structure (Counts / Elements / bearing
    // Allegations) hasn't been loaded into Neo4j — same contract as the
    // causes-of-action endpoint.
    const reason =
      response.status === 404
        ? " — case structure not loaded (run the canonical Element loader)"
        : "";
    throw new Error(
      `Failed to load proof-matrix rollup for "${slug}" (HTTP ${response.status}${reason}). Try reloading the page.`,
    );
  }

  let data: unknown;
  try {
    data = await response.json();
  } catch {
    throw new Error(
      `Proof-matrix rollup response for "${slug}" was not valid JSON (the backend may be down). Try reloading the page.`,
    );
  }

  const parsed = data as Partial<ProofMatrixRollupResponse>;
  if (!Array.isArray(parsed.counts)) {
    throw new Error(
      `Proof-matrix rollup response for "${slug}" is missing the "counts" array — ` +
        `backend/frontend contract mismatch. If reloading does not help, report this to the site administrator.`,
    );
  }

  return parsed as ProofMatrixRollupResponse;
}

/**
 * Re-key the rollup rows into a `count_number → deduped_allegations` lookup so
 * Home can match each Count card to its total in O(1).
 *
 * ## React/TS Learning: a pure data-shaping helper (not business logic)
 * This is a verbatim re-keying of values the endpoint already computed — no
 * summing, no math, no per-Element reasoning (Standing Rule: no business logic
 * in the frontend; the backend's `deduped_allegations` is displayed as-is). It
 * is kept pure (same input → same output, no side effects) so it is trivially
 * unit-testable and stays out of the component, mirroring how Home consumes the
 * `count_descriptions` map.
 *
 * @param counts the rollup rows
 * @returns a record keyed by `count_number` whose values are the deduped totals
 */
export function indexAllegationTotals(
  counts: CountRollup[],
): Record<number, number> {
  const totals: Record<number, number> = {};
  for (const row of counts) {
    totals[row.count_number] = row.deduped_allegations;
  }
  return totals;
}
